spec/controllers/products_controller_spec.rb

```diff
@@ -0,0 +1,198 @@
+    end
+
+    it "renders show view" do
+      get :show, merchant_id: @product.merchant_id, id: @product.id
+      expect(subject).to render_template :show
+    end
+
+    it "finds the average review of a product with a review" do
+      get :show, merchant_id: @product.merchant_id, id: @product.id
+      expect(assigns(:average)).to eq(2)
+    end
+  end
+
+  describe "POST #create" do
+    before(:each) do
+      Merchant.create(username: "Bobbby", email: "email@email.com", password: "password", password_confirmation: "password")
+    end
+    it "redirects to merchant_products path" do
+      post :create, create_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
+    end
+
+    it "renders new template on error" do
+      post :create, bad_params
+      expect(subject).to render_template :new,(@product.merchant_id)
+    end
+  end
+
+  describe 'GET #edit' do
+    it "renders edit template" do
+    merchant.authenticate(session_data)
+    session[:user_id] = merchant.id
+    get :edit, update_params
+    expect(subject).to render_template :edit
+    end
+  end
+
+  describe "PATCH #update" do
+    it "goes to the product index page when successful" do
+      patch :update, update_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
+    end
+    it "renders the edit view when unsuccessful" do
+      patch :update, bad_params
+      expect(subject).to render_template :edit,(@product.id)
+    end
+  end
+
+  describe "DESTROY #destroy" do
+    it "redirects to merchant index page" do
+      delete :destroy, update_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
+      end
+    end
+
+  describe "create a review" do
+    it "redirects to product show page" do
+      post :create_review, good_review
+      expect(Review.count).to eq(1)
+      expect(subject).to redirect_to product_path(1)
+    end
+
+    it "can't be created if merchant tries to review their own product" do
+      # log in merchant
+      merchant.authenticate(session_data)
+      session[:user_id] = merchant.id
+
+      post :create_review, good_review
+      expect(Review.count).to eq(0)
+      expect(subject).to redirect_to product_path(1)
+    end
+
+    it "can be created if merchant reviews a product they don't own" do
+      merchant.products << @product
+
+      # log in merchant
+      merchant2.authenticate(session_data)
+      session[:user_id] = merchant2.id
+
+      post :create_review, good_review
+
+      expect(Review.count).to eq(1)
+      expect(subject).to redirect_to product_path(1)
+    end
+  end
+
+  describe "retire a product" do
+    it "redirects to merchant index page" do
+      merchant.authenticate(session_data)
+      session[:user_id] = merchant.id
+      patch :retire, update_params
+      expect(subject).to redirect_to merchant_products_path(@product.merchant_id)
```

Comment:
Since this controller method also sets the product's status to "Retired", it would be good to test that behavior as well.